File: LeetCode/289 Game of Life/solution.py
from typing import List

class Solution:
    def gameOfLife(self, board: List[List[int]]) -> None:
        m, n = len(board), len(board[0])

        for i in range(m):
            for j in range(n):
                live = 0
                for x in range(max(0, i - 1), min(m, i + 2)):
                    for y in range(max(0, j - 1), min(n, j + 2)):
                        if (x != i or y != j) and board[x][y] in (1, 2):
                            live += 1
                if board[i][j] == 1:
                    if live < 2 or live > 3:
                        board[i][j] = 2
                elif live == 3:
                    board[i][j] = 3

        for i in range(m):
            for j in range(n):
                board[i][j] %= 2

File: LeetCode/289 Game of Life/solution.rs
impl Solution {
    pub fn game_of_life(board: &mut Vec<Vec<i32>>) {
        let m = board.len();
        let n = board[0].len();

        let directions = [
            (-1, -1), (-1, 0), (-1, 1),
            (0, -1),           (0, 1),
            (1, -1),  (1, 0),  (1, 1),
        ];

        for i in 0..m {
            for j in 0..n {
                let mut live_neighbors = 0;

                for &(dx, dy) in &directions {
                    let ni = i as i32 + dx;
                    let nj = j as i32 + dy;

                    if ni >= 0
                        && ni < m as i32
                        && nj >= 0
                        && nj < n as i32
                    {
                        let val = board[ni as usize][nj as usize];
                        if val == 1 || val == 2 {
                            live_neighbors += 1;
                        }
                    }
                }

                if board[i][j] == 1 {
                    if live_neighbors < 2 || live_neighbors > 3 {
                        board[i][j] = 2; // live -> dead
                    }
                } else {
                    if live_neighbors == 3 {
                        board[i][j] = 3; // dead -> live
                    }
                }
            }
        }

        // Finalize the board
        for i in 0..m {
            for j in 0..n {
                board[i][j] = match board[i][j] {
                    2 => 0,
                    3 => 1,
                    x => x,
                };
            }
        }
    }
}